### ISSUE TITLE
CA1508: False Positive when using Equals(null) on classes with overridden equality

### ISSUE THREAD

**Issue by Dave-EMIS** [dotnet/roslyn-analyzers#4536]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `CA1508: Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

I'm seeing reports of 1508 Dead Code when I do `x.Equals(null)` in my unit tests that are testing my own implementation of Equals correctly handles `null`

### Steps To Reproduce

```csharp
class C
{
	public override bool Equals(object obj)
	{
		return 
			obj is null
			||
			base.Equals(obj); //Or whatever
	}

	public override int GetHashCode() => base.GetHashCode();
}
```

In my tests:

```
var actual = new C().Equals(null);

Assert.False(actual);
```

### Expected behavior

No warning

### Actual behavior

Warning on `new C().Equals(null)` been dead code and replaced by `false`.

> warning CA1508: 'new C().Equals(null)' is always 'false'. Remove or refactor the condition(s) to avoid dead code.

## Additional context

There should be checks on this for if the class has its own equality, that someone might have chosen to make null equal to a value.

**Comment by Dave-EMIS** [comment]
Not sure if a separate issue or not, but I am also seeing it on:

```
using (new AssertionScope)
{
}
```

> warning CA1508: 'new AssertionScope()' is never 'null'. Remove or refactor the condition(s) to avoid dead code.

Which is a common pattern in Fluent Assertions

**Comment by Youssef1313** [comment]
> Not sure if a separate issue or not, but I am also seeing it on:
> 
> ```
> using (new AssertionScope)
> {
>   5.Should().Be(6);
> }
> ```
> 
> > warning CA1508: 'new AssertionScope()' is never 'null'. Remove or refactor the condition(s) to avoid dead code.
> 
> Which is a common pattern in Fluent Assertions

IIRC, this one was recently fixed.

**Comment by mavasani** [comment]
@Youssef1313 is correct. Verified no repro with https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers/5.0.1
